Lib/test/test_posix.py

```diff
@@ -1476,6 +1478,131 @@ def test_empty_file_actions(self):
         )
         self.assertEqual(os.waitpid(pid, 0), (pid, 0))
 
+    def test_resetids_explicit_default(self):
+        pid = posix.posix_spawn(
+            sys.executable,
+            [sys.executable, '-c', 'pass'],
+            os.environ,
+            resetids=False
+        )
+        self.assertEqual(os.waitpid(pid, 0), (pid, 0))
+
+    def test_resetids(self):
+        pid = posix.posix_spawn(
+            sys.executable,
+            [sys.executable, '-c', 'pass'],
+            os.environ,
+            resetids=True
+        )
+        self.assertEqual(os.waitpid(pid, 0), (pid, 0))
+
+    def test_resetids_wrong_type(self):
+        with self.assertRaises(TypeError):
+            posix.posix_spawn(sys.executable,
+                              [sys.executable, "-c", "pass"],
+                              os.environ, resetids=0)
+        with self.assertRaises(TypeError):
+            posix.posix_spawn(sys.executable,
+                              [sys.executable, "-c", "pass"],
+                              os.environ, resetids=None)
+
+    def test_setpgroup(self):
+        pid = posix.posix_spawn(
+            sys.executable,
+            [sys.executable, '-c', 'pass'],
+            os.environ,
+            setpgroup=os.getpgrp()
+        )
+        self.assertEqual(os.waitpid(pid, 0), (pid, 0))
+
+    @unittest.skipUnless(hasattr(signal, 'pthread_sigmask'),
+                           'need signal.pthread_sigmask()')
+    def test_setsigmask(self):
+        code = textwrap.dedent("""\
+            import _testcapi, signal
+            _testcapi.raise_signal(signal.SIGUSR1)""")
+
+        pid = posix.posix_spawn(
+            sys.executable,
+            [sys.executable, '-c', code],
+            os.environ,
+            setsigmask=[signal.SIGUSR1]
+        )
+        self.assertEqual(os.waitpid(pid, 0), (pid, 0))
+
+    def test_setsigmask_wrong_type(self):
+        with self.assertRaises(TypeError):
+            posix.posix_spawn(sys.executable,
+                              [sys.executable, "-c", "pass"],
+                              os.environ, setsigmask=34)
+        with self.assertRaises(TypeError):
+            posix.posix_spawn(sys.executable,
+                              [sys.executable, "-c", "pass"],
+                              os.environ, setsigmask=["j"])
+        with self.assertRaises(ValueError):
+            posix.posix_spawn(sys.executable,
+                              [sys.executable, "-c", "pass"],
+                              os.environ, setsigmask=[9998, 9999])
+
+    @unittest.skipUnless(hasattr(posix, 'sched_setscheduler'), "can't change scheduler")
+    def test_setscheduler_only_param(self):
+        pid = posix.posix_spawn(
+            sys.executable,
+            [sys.executable, '-c', "pass"],
```

Comment:
I am not sure this is possible without root as as far as I understand we cannot change the scheduling policy of the parent or the child and therefore we cannot use anything for the scheduler parameters other than the default. Some ideas on how to solve this issue?

Reply:
I concur with you. Seems this can't be tested without root. The test for root can be written, but I'm fine with a simple test for now.

Calling `os.sched_param(os.PRIO_PROCESS)` is incorrect. It works only by accident, because the numeric value of `os.PRIO_PROCESS` matches the current priority (0). You need to pass the value between `os.sched_get_priority_min()` and `os.sched_get_priority_max()`. It is safe to pass `os.sched_get_priority_min(os.sched_getscheduler(0))` and check it inside the child. For example:
```python
policy = os.sched_getscheduler(0)
priority = os.sched_get_priority_min(policy)
code = f"""import os
if os.sched_getscheduler(0) != {policy}:
    os.exit(101)
if os.sched_getparam(0).sched_priority != {priority}:
    os.exit(102)
"""
pid = posix.posix_spawn(..., scheduler=(None, os.sched_param(priority)))
...
# in test_setscheduler_with_policy
pid = posix.posix_spawn(..., scheduler=(policy, os.sched_param(priority)))
```

Reply:
Yup, sorry about that. By the way, notice that in general only `SCHED_FIFO` and `SCHED_RR` are allowed in almost all implementations of `glibc` (example: https://github.com/lattera/glibc/blob/a2f34833b1042d5d8eeb263b4cf4caaea138c4ad/posix/spawnattr_setschedpolicy.c ) should we say this in the docs?

Also, only this options allow priorities different than 0 according to http://man7.org/linux/man-pages/man7/sched.7.html.